File: WKWebView-Print/WKWebView-Print/NSLayoutConstraint+Copy.m
//
//  NSLayoutConstraint+Copy.m
//  WKWebView-Print
//
//  http://github.com/sergemoskalenko/WKWebView-Print-Simple
//  Created by Serge Moskalenko on 19.07.2018.
//  Copyright © 2018 Serge Moskalenko. All rights reserved.
//

#import "NSLayoutConstraint+Copy.h"

@implementation NSLayoutConstraint(Copy)
+(instancetype)constraintWithConstraint:(NSLayoutConstraint *)constraint forView:(UIView *)oldView newView:(UIView *)newView {
    NSLayoutConstraint* newConstraint = [NSLayoutConstraint constraintWithItem:(constraint.firstItem == oldView ? newView : constraint.firstItem) attribute:constraint.firstAttribute relatedBy:constraint.relation toItem:(constraint.secondItem == oldView ? newView : constraint.secondItem) attribute:constraint.secondAttribute multiplier:constraint.multiplier constant:constraint.constant];
    
    return newConstraint;
}

+ (NSArray<__kindof NSLayoutConstraint *> *)constraintsWitView:(UIView *)oldView newView:(UIView *)newView {
    NSMutableArray* newConstraints = [@[] mutableCopy];
    
    for (NSLayoutConstraint* constraint in oldView.constraints) {
        if (constraint.firstItem == oldView || constraint.secondItem == oldView) {
            NSLayoutConstraint* newConstraint = [self constraintWithConstraint:constraint forView:oldView newView:newView];
            if (newConstraint)
                [newConstraints addObject:newConstraint];
        }
    }
    
    return [newConstraints copy];
}

+ (NSArray<__kindof NSLayoutConstraint *> *)constraintsWitSuperView:(UIView *)oldView newView:(UIView *)newView {
    NSMutableArray* newConstraints = [@[] mutableCopy];
    
    for (NSLayoutConstraint* constraint in (oldView.superview).constraints) {
        if (constraint.firstItem == oldView || constraint.secondItem == oldView) {
            NSLayoutConstraint* newConstraint = [self constraintWithConstraint:constraint forView:oldView newView:newView];
            if (newConstraint)
                [newConstraints addObject:newConstraint];
        }
    }
    
    return [newConstraints copy];
}

+ (void)copyConstraintsFromView:(UIView *)oldView toView:(UIView *)newView {
    [newView addConstraints:[self constraintsWitView:oldView newView:newView]];
    [newView.superview addConstraints:[self constraintsWitSuperView:oldView newView:newView]];
}

@end

File: WKWebView-Print-Swift/WKWebView-Print-Swift/NSLayoutConstraint+Copy.swift

//
//  NSLayoutConstraint+Copy.m
//  WKWebView-Print
//
//  http://github.com/sergemoskalenko/WKWebView-Print-Simple
//  Created by Serge Moskalenko on 19.07.2018.
//  Copyright © 2018 Serge Moskalenko. All rights reserved.
//

import UIKit

extension NSLayoutConstraint {
    class func copy(constraint: NSLayoutConstraint?, for oldView: UIView?, newView: UIView?)  -> NSLayoutConstraint?  {
        var anItem = constraint?.firstItem
        if let testView:UIView = anItem as! UIView? {
            if testView === oldView {
                anItem = newView
            }
        }
        var anItem2 = constraint?.secondItem
        if let testView:UIView = anItem2 as! UIView? {
            if testView === oldView {
                anItem2 = newView
            }
        }
        let anAttribute = constraint?.firstAttribute
        let aRelation = constraint?.relation
        let anAttribute1 = constraint?.secondAttribute
        let newConstraint: NSLayoutConstraint?  = NSLayoutConstraint(item: anItem as Any, attribute: anAttribute!, relatedBy: aRelation!, toItem: anItem2, attribute: anAttribute1!, multiplier: constraint?.multiplier ?? 0.0, constant: constraint?.constant ?? 0.0)

        return newConstraint
    }
    
    
    class func constraintsWitView(_ oldView: UIView?, newView: UIView?) -> [NSLayoutConstraint]? {
        var newConstraints: [AnyHashable] = []
        for constraint: NSLayoutConstraint? in oldView?.constraints ?? [NSLayoutConstraint?]() {
            
            let anItem = constraint?.firstItem as AnyObject?
            if anItem === oldView as AnyObject {
            } else {
                let anItem2 = constraint?.secondItem  as AnyObject?
                if anItem2 === oldView as AnyObject  {
                } else {
                    continue
                }
            }

            if let newConstraint: NSLayoutConstraint = self.copy(constraint: constraint, for: oldView, newView: newView) {
                newConstraints.append(newConstraint)
            }
        }
        return newConstraints as? [NSLayoutConstraint]
    }
    
    class func constraintsWitSuperView(_ oldView: UIView?, newView: UIView?) -> [NSLayoutConstraint]? {
        var newConstraints: [AnyHashable] = []
        for constraint: NSLayoutConstraint? in (oldView?.superview)?.constraints ?? [NSLayoutConstraint?]() {
            
            let anItem = constraint?.firstItem as AnyObject?
            if anItem === oldView as AnyObject {
            } else {
                let anItem2 = constraint?.secondItem  as AnyObject?
                if anItem2 === oldView as AnyObject  {
                } else {
                    continue
                }
            }
            
            if let newConstraint: NSLayoutConstraint = self.copy(constraint: constraint, for: oldView, newView: newView) {
                newConstraints.append(newConstraint)
            }
        }
        return newConstraints as? [NSLayoutConstraint]
    }
    
    class func copyConstraints(from oldView: UIView?, to newView: UIView?) {
        if let aView = self.constraintsWitView(oldView, newView: newView) {
            newView?.addConstraints(aView)
        }
        if let aView = self.constraintsWitSuperView(oldView, newView: newView) {
            newView?.superview?.addConstraints(aView)
        }
    }
}
